bcarterterp/Flap
# 4-Create App Flavors for Dev and Prod (flutter_flavor)

### **Summary:** 
This PR adds in 2 app flavors. One is called "dev" and can be re-purposed for setting-up a staging api. One is called "prod" and can be re-purposed to set up a production api. 

### **How to run and test this:**
There are 2 ways to run flavors. 

**1) In VSCode**, if you click on the Run and Debug pane, you'll see a drop-down with different build configurations. You can select "dev" and "prod" flavors via this drop-down. 
<img width="430" alt="Screenshot 2023-10-20 at 3 48 48 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/661d8829-f7c4-4435-b538-0bddcad6d482">

**2. In the terminal**, you can run the below commands:

**For Prod:**
`flutter run --flavor prod -t lib/main_prod.dart`

**For Dev:**
`flutter run --flavor dev -t lib/main_dev.dart`

If you run them successfully, you should see the below UI:

**For Prod: (you should see the red banner in top right say prod)**
<img width="246" alt="Screenshot 2023-10-20 at 3 51 59 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/cd5b93c7-5fb6-4e3f-8d73-e746527fd2af">


**For Dev: (you should see the red banner in top right say dev)**
<img width="246" alt="Screenshot 2023-10-20 at 3 55 47 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/0a037c90-d9d0-4102-aa1f-7154b50c1b25">


### **Steps taken to Implement this**

This was implemented following guidelines from the below articles:
https://itnext.io/flutter-new-app-setup-with-flavors-in-one-go-331471b127e3  (great resource for setting up in iOS and Android)
https://pub.dev/packages/flutter_flavor#flutter_flavor (how to set up FlavorConfig in dart)
https://medium.com/@devrient.jonas/flutter-profiles-in-vscode-product-flavors-c95231818005 (setting up flavor profiles to be launched in VS Code)


**Step 1) Add build flavors in Android**
Add flavors to app/build.gradle. The resValue can be used to define any type of resource, but in the case below its the app name that is also defined in the manifest. 
<img width="425" alt="Screenshot 2023-10-20 at 4 44 17 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/909696b4-7769-4203-9e1e-b8775e627e5e">


**Step 2) Add build flavors to iOS**

My learning here is that this requires more steps than android and also relies more on navigating XCode UI (not a quick text edit of a file). In summary, it required creating a new scheme, duplicating build configurations and renaming (ex. "Debug-dev", "Release-dev", "Profile-dev"). Connecting dev build configurations to the dev scheme and prod configurations to the prod scheme. 

This article was the best to follow along because of the gifs showing how to navigate XCode: https://itnext.io/flutter-new-app-setup-with-flavors-in-one-go-331471b127e3

**Step 3) Create 2 entry-points into the app: main_dev.dart and main_prod.dart**
My learning also is that there is no official documentation from Flutter on how to best configure dart code to provide multiple flavors. The flutter docs here only describe setting it up natively (https://docs.flutter.dev/deployment/flavors).

I used flutter_flavors package and followed along with their code descriptions. 
<img width="585" alt="Screenshot 2023-10-20 at 4 38 59 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/36fff471-440a-4a63-ba1e-0c487c84cce1">

**Step 4) Add configurations to launch.json to allow Visual Studio to build the different flavors.** (This is optional - you can always just build from the terminal too but this is easier). 
<img width="340" alt="Screenshot 2023-10-20 at 4 40 26 PM" src="https://github.com/bcarterterp/Flap/assets/14029158/afec0c6e-2414-42ad-906b-cbb2696b54fb">
